### PR TITLE
[SYCL] Fix issues regarding deprecated SYCL 1.2.1 exception types

### PR DESCRIPTION
**Pull request by HPS-1**
Some exception types in SYCL 1.2.1 no longer exist in SYCL 2020 spec but they are marked as "SYCL-2020 deprecated" in our code with a large quantity of usages. They should be replaced with sycl::Exception. For example:
`throw runtime_error("--ERROR MESSAGE--", PI_ERROR_INVALID_DEVICE);`
should be replaced with
`throw sycl::exception(sycl::errc::runtime, "--ERROR MESSAGE--");`

In this PR: change deprecation type macros for deprecated exception types and replace the first batch of sycl::runtime_error instances with sycl::exception.

List of such exception types:
runtime_error
kernel_error
accessor_error
nd_range_error
event_error
invalid_parameter_error
device_error
compile_program_error
link_program_error
invalid_object_error
memory_allocation_error
platform_error
profiling_error
feature_not_supported

### CODE REVIEW

File: sycl/include/sycl/exception.hpp [hunk 1]
```diff
@@ -136,9 +136,8 @@ class __SYCL_EXPORT exception : public virtual std::exception {
             const char *WhatArg);
 };
 
-class __SYCL2020_DEPRECATED(
-    "use sycl::exception with sycl::errc::runtime instead.") runtime_error
-    : public exception {
+class __SYCL_DEPRECATED("use sycl::exception with sycl::errc::runtime instead.")
```
Comment (AlexeySachkov) [inline sycl/include/sycl/exception.hpp:139]: Strictly speaking, this is an incorrect change. Those classes are still valid in SYCL 1.2.1 mode which can still be selected with our implementation:

https://github.com/intel/llvm/blob/c821dc934dc7934b0209b5d3f88a280bbaa7145c/clang/include/clang/Driver/Options.td#L8316-L8321

I'm sure that it has been completely broken for a while now, but I would prefer to have a separate PR that completely removes it consistently from everywhere

**Review by AlexeySachkov — COMMENTED** [review]
LGTM except for this one comment

**Review by gmlueck — APPROVED** [review]
Spec changes LGTM.

File: sycl/include/sycl/ext/oneapi/backend/hip.hpp [hunk 2]
```diff
@@ -16,10 +16,8 @@ inline namespace _V1 {
 template <>
 inline backend_return_t<backend::ext_oneapi_hip, device>
 get_native<backend::ext_oneapi_hip, device>(const device &Obj) {
-  // TODO swap with SYCL 2020 exception when in ABI-break window
   if (Obj.get_backend() != backend::ext_oneapi_hip) {
-    throw sycl::runtime_error(errc::backend_mismatch, "Backends mismatch",
-                              PI_ERROR_INVALID_OPERATION);
+    throw sycl::exception(errc::backend_mismatch, "Backends mismatch");
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/ext/oneapi/backend/hip.hpp:20]: [ABI-break] tag is missing?

**Review by dm-vodopyanov — APPROVED** [review]

**Pull request closed without merging**